### PR TITLE
Troyfeng116/multi doc

### PR DESCRIPTION
**Pull request by troyfeng116**
- Add backend-flexible `MultiDocJointModel` and `MultiDocSeparateModel` for multi-document summarization

### CODE REVIEW

File: model/multi_doc/base_model.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+from model.base_model import SummModel
```
Comment (niansong1996) [inline model/multi_doc/base_model.py:1]: I think it's better to call this file smth like `multi_doc_base_model.py`, so we don't have to use parent dir to determine which is which?

File: model/multi_doc/base_model.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    def __init__(self,
+                 trained_domain: str = None,
+                 max_input_length: int = None,
+                 max_output_length: int = None,
+                 ):
+        super(MultiDocSummModel, self).__init__(trained_domain = trained_domain, max_input_length = max_input_length, max_output_length = max_output_length)
+
+    @classmethod
+    def assert_summ_input_type(cls, corpus, query):
+        for instance in corpus:
```
Comment (niansong1996) [inline model/multi_doc/base_model.py:15]: you don't have to use `for` loop here, you can do `if not all([isinstance(ins, list) for ins in corpus]): ...`

similarly, you can drop the nested loops and do `if not all([isinstance(doc, str) for doc in ins for ins in corpus])

File: model/multi_doc/multi_doc_joint_model.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        self.assert_summ_input_type(corpus, None)
+        joint_corpus = []
+        for instance in corpus:
+            joint_corpus.append([" ".join(instance)])
+
+        summaries = []
+        for joint_multi_doc_instance in joint_corpus:
+            summary = self.model.summarize(joint_multi_doc_instance)
+            summaries.append(summary[0])
+
```
Comment (niansong1996) [inline model/multi_doc/multi_doc_joint_model.py:24]: L14-23 looks strange, why not make `joint_corpus` a `List[str]` and just call `self.model.summarize()` only once?

File: model/multi_doc/base_model.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            if not isinstance(instance, list):
+                raise TypeError("Multi-document summarization models summarize instances of multiple documents (`List[List[str]]`).")
+            for doc in instance:
+                if type(doc) != str:
+                    raise TypeError("Multi-document summarization models summarize instances of multiple documents (`List[List[str]]`).")
+
+        if query is not None:
+            if not isinstance(query, list):
```
Comment (niansong1996) [inline model/multi_doc/base_model.py:23]: The same `all()` checker comment goes here

File: model/single_doc/base_model.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+class SingleDocSummModel(SummModel):
+    def __init__(self,
+                 trained_domain: str = None,
+                 max_input_length: int = None,
+                 max_output_length: int = None,
+                 ):
+        super(SingleDocSummModel, self).__init__(trained_domain = trained_domain, max_input_length = max_input_length, max_output_length = max_output_length)
+
+    @classmethod
+    def assert_summ_input_type(cls, corpus, query):
```
Comment (niansong1996) [inline model/single_doc/base_model.py:14]: The same `all()` checker comment goes here

**Review by niansong1996 — COMMENTED** [review]
Made some comments, mostly about a better way to do type checking. Otherwise, very solid work!

**Comment by niansong1996** [comment]
LGTM. @troyfeng116  Are all the tests still passing? If so, I think we can merge this

**Comment by troyfeng116** [comment]
> LGTM. @troyfeng116 Are all the tests still passing? If so, I think we can merge this

Yep, all tests passing

**Pull request merged**